app/models/upload_reference.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+class UploadReference < ActiveRecord::Base
+  belongs_to :upload
+  belongs_to :target, polymorphic: true
+
+  def self.ensure_exist!(upload_ids: [], target: nil, target_type: nil, target_id: nil)
+    raise "target OR target_type and target_id are required" if !target && !(target_type && target_id)
+
+    if target.present?
+      target_type = target.class
+      target_id = target.id
+    end
+
+    upload_ids = upload_ids.uniq.reject(&:blank?)
+    target_type = target_type.to_s
+
+    if upload_ids.empty?
+      UploadReference
```

Comment:
With this I feel like the `ensure_exists!` name is a little deceiving, because in this case it's deleting something. Maybe a `sync!` would be better? Or perhaps `upsert!`?

Reply:
It used to be called just `exists!` and Regis suggested this name, which I like better. `upsert!` is also wrong because as you said, it is not just `update + insert`, but also `delete`. `sync!` sounds fine too.